Add one end-to-end functional test

pcraig3 opened the pull request:
Adding [cypress] to our test dependencies and adding one end-to-end test that goes from the welcome page through to the confirmation page.

Cypress can be run in the browser and then stepped through using their visual interface (`npm run cypress`), or cypress can be run on the command line for text-only test output (`npm run cypress:cli`). The `cli` command means that we can incorporate this into our GitHub Actions workflow, so that they will run on every open PR. As we build out more complex flows, we're going to want more than just unit tests to keep us confident that we're not regressing.

- [x] add one end to end functional test
- [x] run in a browser or on the command line
- [x] incorporate into github actions
- [x] create user fixture rather than test fixture

To run this locally:

```bash
git pull
git checkout cypress
npm install
npm run cypress  # to run in a browser
npm run cypress:cli  # to run on the command line
```

### Cypress using a browser

![cypress](https://user-images.githubusercontent.com/2454380/56380493-49309500-61e0-11e9-9102-336e2a0551b1.gif)

### Cypress on the command line

![cypress-cli](https://user-images.githubusercontent.com/2454380/56380510-53eb2a00-61e0-11e9-91e0-1a6961b54202.gif)

pcraig3 committed:
Add cypress as a test dependency

Cypress runs functional tests in a pretty reasonable way.

Means we can test out the main flows in our app and be confident
that they continue to work as we add new code.

Docs are pretty good too:
- https://docs.cypress.io/guides/getting-started/writing-your-first-test.html

pcraig3 committed:
First test

Just loads the home page but doesn't do anything else.

pcraig3 committed:
Ignore some cypress folders that we're not using

It looks like Cypress (re)creates a default directory structure
when first initiated, but we don't need a bunch of placeholder
code sitting around if it's not being used.

We can always add it back in later if we need to.

pcraig3 committed:
Add a full_run functional test

We can test for other non-happy path cases as well, but at minimum
we want to have a full run through the app.

(We can update this as we go.)

pcraig3 committed:
Add 'cypress run' cli comment

Running 'cypress open' will pop open a browser and let you see the
actual tests as they move through the interface.

However, if we want to run these tests more quickly, and, particularly,
as part of our CD pipeline, we need a trimmer way to run our functional
tests.

So this should be able to run inside of GitHub actions.

pcraig3 committed:
Add cypress functional test run to github actions

Also reformatted the file because I have a github actions
VSCode extension which wants to see it spaced a particular way.

pcraig3 committed:
Update workflow with base container that can run cypress

pcraig3 committed:
Add user fixture to cypress test

Main difference is that on the dashboard, we can check for all the
expected values without hardcoding them.

sboisvert commented on cypress/fixtures/user.json:
```diff
@@ -0,0 +1,7 @@
+  "address": "850 Argyle Street\nPort Alberni\nBritish Columbia\nV9Y 1V8",
+  "maritalStatus": "Married",
+  "children": "0",
+  "income": "$25,000"
```
Should we change this in this PR to match the new T4 flow?

pcraig3 replied:
Yeah, I can update it. I think let's wait for #28 to make it in.

sboisvert commented on cypress/integration/full_run.spec.js:
```diff
@@ -0,0 +1,50 @@
+
+      cy.get('a.buttonLink')
+        .should('contain', 'Get started')
+        .click()
```
I think this would also break in the new flow.

pcraig3 committed:
Update dependencies

Only one dependency actually.
We're just using it for testing and it wasn't a breaking change.

pcraig3 committed:
Update user fixture and add new pages

Lots of copy and paste in here, so I'll trim it down a bit

pcraig3 committed:
Refactor checking summary table rows

Created a function to check row names and values so that we have less
boilerplate cypress code floating around up in here.
